File: backend/app/schemas/replay.py
"""API contracts for the historical CSV playback, separate from live NDTP."""

from datetime import datetime

from pydantic import BaseModel


class ReplayVehicle(BaseModel):
    tr_id: int
    unit_id: int
    samples: int
    start_at: datetime
    end_at: datetime


class ReplayTelemetry(BaseModel):
    available_at: datetime
    event_time: datetime
    lat: float
    lon: float
    speed: float | None
    heading: float | None = None


class ReplayStop(BaseModel):
    stop_id: int
    planned_at: datetime
    lat: float
    lon: float
    place_id: str = ""
    address: str | None = None
    gps_confirmed: bool = False
    gps_distance_m: float | None = None


class ReplayPoint(BaseModel):
    sample_id: str
    T: datetime
    target_stop_id: int
    target_time_begin: datetime
    cur_dev_s: float
    actual_delay_s: float
    actual_at: datetime
    lat: float
    lon: float


class ReplayScenario(BaseModel):
    source: str = "historical_csv_test"
    vehicle: ReplayVehicle
    telemetry: list[ReplayTelemetry]
    stops: list[ReplayStop]
    points: list[ReplayPoint]


class ReplayPrediction(BaseModel):
    sample_id: str
    predicted_delay_s: float
    baseline_delay_s: float
    model_version: str


class ReplayRun(BaseModel):
    run_id: str
    tr_id: int
    start_at: datetime
    end_at: datetime
    stop_ids: list[int]
    confirmed_stops: int
    valid: bool


class ReplayFleetVehicle(BaseModel):
    tr_id: int
    unit_id: int
    telemetry: list[ReplayTelemetry]
    stops: list[ReplayStop]
    runs: list[ReplayRun]
    points: list[ReplayPoint]


class ReplayFleet(BaseModel):
    source: str = "historical_csv_test"
    start_at: datetime
    end_at: datetime
    vehicles: list[ReplayFleetVehicle]


class ReplayStreamStatus(BaseModel):
    state: str
    source_at: datetime | None
    time_mode: str | None
    speed_multiplier: float
    sent_packets: int
    total_packets: int

File: frontend/src/types/replay.ts
export type ReplayVehicle = {
  tr_id: number
  unit_id: number
  samples: number
  start_at: string
  end_at: string
}

export type ReplayTelemetry = {
  available_at: string
  event_time: string
  lat: number
  lon: number
  speed: number | null
  heading: number | null
}

export type ReplayStop = {
  stop_id: number
  planned_at: string
  lat: number
  lon: number
  place_id: string
  address: string | null
  gps_confirmed: boolean
  gps_distance_m: number | null
}

export type ReplayPoint = {
  sample_id: string
  T: string
  target_stop_id: number
  target_time_begin: string
  cur_dev_s: number
  actual_delay_s: number
  actual_at: string
  lat: number
  lon: number
}

export type ReplayScenario = {
  source: 'historical_csv_test'
  vehicle: ReplayVehicle
  telemetry: ReplayTelemetry[]
  stops: ReplayStop[]
  points: ReplayPoint[]
}

export type ReplayPrediction = {
  sample_id: string
  predicted_delay_s: number
  baseline_delay_s: number
  model_version: string
}

export type ReplayRun = {
  run_id: string
  tr_id: number
  start_at: string
  end_at: string
  stop_ids: number[]
  confirmed_stops: number
  valid: boolean
}

export type ReplayFleetVehicle = {
  tr_id: number
  unit_id: number
  telemetry: ReplayTelemetry[]
  stops: ReplayStop[]
  runs: ReplayRun[]
  points: ReplayPoint[]
}

export type ReplayFleet = {
  source: 'historical_csv_test'
  start_at: string
  end_at: string
  vehicles: ReplayFleetVehicle[]
}

export type ReplayStreamStatus = {
  state: string
  source_at: string | null
  time_mode: string | null
  speed_multiplier: number
  sent_packets: number
  total_packets: number
}
